hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.hadoop.hdds.scm.container.ContainerID;
+import org.apache.hadoop.hdds.scm.container.ContainerInfo;
+import org.apache.hadoop.hdds.scm.container.ContainerManagerV2;
+import org.apache.hadoop.hdds.scm.container.ContainerNotFoundException;
+import org.apache.hadoop.hdds.scm.container.ContainerReplica;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * Find a target giving preference to more under-utilized nodes.
+ */
+public class FindTargetGreedy implements FindTargetStrategy {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(FindTargetGreedy.class);
+
+  private ContainerManagerV2 containerManager;
+  private PlacementPolicy placementPolicy;
+
+  public FindTargetGreedy(
+      ContainerManagerV2 containerManager,
+      PlacementPolicy placementPolicy) {
+    this.containerManager = containerManager;
+    this.placementPolicy = placementPolicy;
+  }
+
+  /**
+   * Find a {@link ContainerMoveSelection} consisting of a target and
+   * container to move for a source datanode. Favours more under-utilized nodes.
+   * @param source Datanode to find a target for
+   * @param potentialTargets Collection of potential target datanodes
+   * @param candidateContainers Set of candidate containers satisfying
+   *                            selection criteria
+   *                            {@link ContainerBalancerSelectionCriteria}
+   * @param canSizeEnterTarget A functional interface whose apply
+   * (DatanodeDetails, Long) method returns true if the size specified in the
+   * second argument can enter the specified DatanodeDetails node
+   * @return Found target and container
+   */
+  @Override
+  public ContainerMoveSelection findTargetForContainerMove(
+      DatanodeDetails source, Collection<DatanodeDetails> potentialTargets,
+      Set<ContainerID> candidateContainers,
+      BiFunction<DatanodeDetails, Long, Boolean> canSizeEnterTarget) {
+    for (DatanodeDetails target : potentialTargets) {
+      for (ContainerID container : candidateContainers) {
+        Set<ContainerReplica> replicas;
+        ContainerInfo containerInfo;
+
+        try {
+          replicas = containerManager.getContainerReplicas(container);
+          containerInfo = containerManager.getContainer(container);
+        } catch (ContainerNotFoundException e) {
+          LOG.warn("Could not get Container {} from Container Manager for " +
+              "obtaining replicas in Container Balancer.", container, e);
+          continue;
+        }
+
+        if (replicas.stream().noneMatch(
+            replica -> replica.getDatanodeDetails().equals(target)) &&
+            containerMoveSatisfiesPlacementPolicy(container, replicas, source,
+            target) &&
+            canSizeEnterTarget.apply(target, containerInfo.getUsedBytes())) {
```

Comment:
maybe later we can take networkTopology in to account,  for all the candidate targets , we always choose the nearest one in networkTopology. this can be done in a new jira, maybe i can help doing this

Reply:
Yes, we can do this in another jira.